exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/ComplexUnnestVisitor.java

```diff
@@ -0,0 +1,195 @@
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Visitor that moves non-{@link RexFieldAccess} rex node from project below {@link Uncollect}
+ * to the left side of the {@link Correlate}.
+ */
+public class ComplexUnnestVisitor extends RelShuttleImpl {
+  private static final String COMPLEX_FIELD_NAME = "$COMPLEX_FIELD_NAME";
+
+  private final Map<CorrelationId, RelNode> leftInputs = new HashMap<>();
+  private final Map<CorrelationId, CorrelationId> updatedCorrelationIds = new HashMap<>();
+
+  private ComplexUnnestVisitor() {
+  }
+
+  @Override
+  public RelNode visit(LogicalCorrelate correlate) {
+    RelNode left = correlate.getLeft().accept(this);
+    leftInputs.put(correlate.getCorrelationId(), left);
+
+    RelNode right = correlate.getRight().accept(this);
+    // if right input wasn't changed or left input wasn't changed
+    // after rewriting right input, no need to create Correlate with new CorrelationId
+    if (correlate.getRight() == right
+        || left == leftInputs.get(correlate.getCorrelationId())) {
+      if (correlate.getLeft() == left) {
+        return correlate;
+      }
+      // changed only inputs, but CorrelationId left the same
+      return correlate.copy(correlate.getTraitSet(), Arrays.asList(left, right));
+    }
+
+    Correlate newCorrelate = correlate.copy(correlate.getTraitSet(),
+        leftInputs.get(correlate.getCorrelationId()), right,
+        updatedCorrelationIds.get(correlate.getCorrelationId()),
+        ImmutableBitSet.of(left.getRowType().getFieldCount()), correlate.getJoinType());
+
+    RelBuilder builder = DrillRelFactories.LOGICAL_BUILDER.create(correlate.getCluster(), null);
+    builder.push(newCorrelate);
+
+    List<RexNode> topProjectExpressions = left.getRowType().getFieldList().stream()
+        .map(field -> builder.getRexBuilder().makeInputRef(left, field.getIndex()))
+        .collect(Collectors.toList());
+
+    switch (correlate.getJoinType()) {
+      case LEFT:
+      case INNER:
+        // adds field from the right input of correlate to the top project
+        topProjectExpressions.add(
+            builder.getRexBuilder().makeInputRef(newCorrelate, topProjectExpressions.size() + 1));
+        // fall through
+      case ANTI:
+      case SEMI:
+        builder.project(topProjectExpressions, correlate.getRowType().getFieldNames());
+    }
+    return builder.build();
+  }
+
+  @Override
+  public RelNode visit(RelNode other) {
+    if (other instanceof Uncollect) {
+      return visit((Uncollect) other);
+    }
+    return super.visit(other);
+  }
+
+  public RelNode visit(Uncollect uncollect) {
+    RelBuilder builder = DrillRelFactories.LOGICAL_BUILDER.create(uncollect.getCluster(), null);
+    RexBuilder rexBuilder = builder.getRexBuilder();
+
+    assert uncollect.getInput() instanceof Project : "Uncollect should have Project input";
```

Comment:
why is this assert required? Is it better to have a if clause and return instead of assert?

Reply:
In the case, if at this point we have something that differs from Project, then something went wrong.